documentation/Sin Cos.tex

```diff
@@ -918,4 +919,67 @@ \subsubsection*{Cos}\label{secerroranalysiscos}
 \]
 This function reaches its extrema on the corners of the trapezoidal domain and is found to be smaller than $2^{-74.348}$ for all $k$, so its contribution to the overall relative error would be very small.  (The term for $n = 1$, on the other hand, induces an error of the order of $2^{-63.346}$ so it must be computed.)
 
+\section*{Annex: Proof of the FMA Computation}\label{prooffma}
+
+\cite{StehléZimmermann2005} presents, in section 2.1, the following algorithm to compute $s_k + c_k h$ exactly:
+\[
+\begin{dcases}
+h' &= \round{s_k + c_k h} \\
+t &= h' - s_k \\
+l &= \round{t - c_k h}
+\end{dcases}
+\]
+where the computation of $t$ is exact if $h'$ and $s_k$ verify the conditions of Sterbenz's Lemma.  \cite{StehléZimmermann2005} then states without proof that $h' + l = s_k + c_k h$.
+
+The first thing to note is that there is a sign error in that last formula.  Let's make the absolute errors $\gz_i$ of each computation explicit:
+\[
+\begin{dcases}
+h' &= s_k + c_k h + \gz_1 \\
+t &= h' - s_k = c_k h + \gz_1 \\
+l &= \round{t - c_k h} = \round{\gz_1} = \gz_1 + \gz_2
+\end{dcases}
+\]
+It is clear that the sum $h' + l = s_k + c_k h + 2 \; \gz_1 + \gz_2$ doubles the error $\gz_1$ (which is large) while the difference $h' - l = s_k + c_k h - \gz_2$ eliminates it and only retains $\gz_2$ (which is much smaller).
+
+To prove that $h' - l = s_k + c_k h$ we must prove that $\gz_2 = 0$, which is equivalent to $\round{\gz_1} = \gz_1$.  In other words, if $\gz_1$ is a machine number then $\gz_2 = 0$ and the equality holds.
+
+The first thing that we show is that the exact value $s_k + c_k h$ has at most $2 M + 1$ bits in its binary representation.  Note that $h = x - x_k$, where the subtraction is exact, and therefore there exists $N \in \intclop{0}{2^M}$ such that $h = N \ULP\of{x_k}$.  Moreover, because $x_k \in \intclos{0}{\frac{\Pi}{4}}$ we have $1/2 < c_k < 1$ and therefore $\ULP\of{c_k} = \ULP\of{1/2} = 2^{-M}$.  Finally, $x_k / 2 < s_k < x_k$, which implies that $\ULP\of{s_k}$ is either $\ULP\of{x_k}$ or $\ULP\of{x_k} / 2$.  We need to distinguish two cases:
```

Comment:
```suggestion
The first thing that we show is that the exact value $s_k + c_k h$ has at most $2 M + 1$ bits in its binary representation.  Note that $h = x - x_k$, where the subtraction is exact, and therefore there exists $N \in \intclop{0}{2^M} \Intersection \Naturals$ such that $h = N \ULP\of{x_k}$.  Moreover, because $x_k \in \intclos{0}{\frac{\Pi}{4}}$ we have $1/2 < c_k < 1$ and therefore $\ULP\of{c_k} = \ULP\of{1/2} = 2^{-M}$.  Finally, $x_k / 2 < s_k < x_k$, which implies that $\ULP\of{s_k}$ is either $\ULP\of{x_k}$ or $\ULP\of{x_k} / 2$.  We need to distinguish two cases:
```